### ISSUE TITLE
[XAI] SaCas9 XAI Comparison Pipeline: Steps 2-5 (Integrated Gradients, ISM/Mismatch, Model B Validation, Case Study Export)

### ISSUE THREAD

**Issue by shin584** [shin584/project#9]
## Problem Statement

Model B (the NT-embedding + physical-feature XGBoost fusion model) beats Model A (the CNN+RNN model) on aggregate metrics (Spearman ρ: 0.7661 → 0.7981, MSE: 0.0231 → 0.0196), but an aggregate metric gain alone doesn't tell us *why*. Without deeper analysis we can't tell whether Model B improved because it learned real SaCas9 biological mechanism (PAM/seed-region sensitivity, thermodynamic stability effects) or because it merely fits noise/dataset artifacts better than Model A did.

This matters for two downstream consumers who currently have nothing to work from:
- Model tuning (feedback loop "A"): without knowing *which* signal (sequence embedding vs. physical features) drives Model B's gains, there's no evidence base for future architecture/feature decisions.
- The XAI demo dashboard (project "C"): it needs a stable, queryable database of per-sample explainability artifacts (attributions, SHAP values, attention maps, case studies) to display — none exists yet.

Step 1 of the analysis plan (Testset isolation, `Model_A_Predictor`, `Model_B_Predictor`, baseline metric verification) is already built and verified (`test_dataset_verifier.py`, `model_a_wrapper.py`, `model_b_wrapper.py`, `step1_verify.py`). Steps 2–5 — the actual explainability and mechanism-validation analysis — are fully specified in `plan_realize.md` but not yet implemented.

## Solution

Implement Steps 2–5 of the finalized analysis pipeline (`plan_realize.md`) against the fixed, SHA256-verified 514-sample Testset, producing a standardized artifact pair (`model_analysis_arrays.npz` + `model_analysis_summary.json`) that both the model-tuning feedback loop and the XAI dashboard can consume:

- **Step 2 — Sequence-level backbone comparison (Integrated Gradients):** compare how sharply Model A vs. Model B's NT regression head attribute importance to PAM (`NNGRRT`) and Seed-region (1–8bp) positions.
- **Step 3 — Biological perturbation simulation (ISM & Mismatch):** in-silico saturation mutagenesis and mismatch-tolerance profiling, run through both models' `predict()`, to see whether either model's sensitivity pattern tracks known SaCas9 binding biology.
- **Step 4 — Model B's own hybrid-fusion validation:** ablation (embedding-only vs. physical-only vs. full), Tree SHAP (global + per-physical-feature), SHAP dependence plots, and Attention Rollout — establishing whether the physical features (not just the NT embedding) are doing real work, and what the NT backbone's internal token-dependency structure looks like.
- **Step 5 — Case Study selection + export:** select the 10 representative samples (Primary/Secondary Discordant + Concordant, per `CONTEXT.md`'s definitions), attach per-sample SHAP/IG/attention detail, and export everything to the standard schema.

## User Stories

1. As a model developer, I want per-position Integrated Gradients attribution for both Model A and Model B on the same Testset sequences, so that I can see whether Model B's NT backbone resolves PAM/Seed positions more sharply than Model A's CNN+RNN backbone.
2. As a model developer, I want Model B's Integrated Gradients computed against the NT Phase 4 regression checkpoint's own differentiable head (not through the XGBoost tree), so that the attribution is mathematically valid (trees aren't differentiable).
3. As a model developer, I want the Model B Integrated Gradients baseline forced to a validated Zero Embedding Vector and the embedding tensor shape asserted before attribution runs, so that a silent shape mismatch never produces misleading attributions.
4. As a model developer, I want 6-mer token attributions projected back to individual nucleotides using the documented Deterministic Projection Rule (even 6-way split), so that 36bp-level comparisons with Model A's per-nucleotide attributions are possible, with the assumption clearly labeled as an analysis convention rather than a claim about the model's internal representation.
5. As a model developer, I want attribution compared via within-sequence L1-normalized relative distribution rather than raw magnitude, so that Model A and Model B (different architectures, different attribution scales) are compared on a like-for-like basis.
6. As a model developer, I want an in-silico saturation mutagenesis (ISM) sweep of every position x every alternative base for both models, so that I can build a position-sensitivity heatmap and check whether Model B degrades cleavage-efficiency predictions more sharply at PAM/Seed positions than Model A does.
7. As a model developer, I want Model B's physical features (MFE, ΔG, Tm, GC) recomputed per-mutant during ISM (not reused from the wild-type sequence), so that the physical-feature contribution to a mutant's score reflects that mutant's actual sequence.
8. As a model developer, I want ISM's relative ΔScore calculation guarded against wild-type scores near zero (|Score_WT| <= 1e-4 masked to NaN), so that a near-zero denominator doesn't produce a spurious, unbounded relative change.
9. As a model developer, I want alternative bases in ISM/mismatch arrays indexed in a fixed lexicographic order (sorted {A,C,G,T} minus wild-type), so that array position 0/1/2 means the same alternative base across every sample and script.
10. As a model developer, I want a mismatch-tolerance profile that tracks predicted efficiency decay from Seed to Distal positions for both models, so that I can compare how well each model's degradation curve matches known PAM-proximal binding-disruption biology.
11. As a model developer, I want single-mismatch results reported two ways — the full 3-alternative-base delta array, and a single worst-case-per-position summary — so that downstream consumers get both full traceability and one unambiguous "how bad can it get here" number without axis confusion.
12. As a model developer, I want a fixed set of 15 complex mismatch scenarios (5 Seed + 5 Distal + 5 Intermittent, seed=42), each carrying its own mutation metadata (positions, ref/alt bases, region label), so that every consumer of `mismatch_complex` sees the same reproducible scenarios with self-describing provenance.
13. As a model developer, I want three independently retrained XGBoost variants — Model B-Embedding, Model B-Physical, Model B-Full — trained on the same dev split, seed, and hyperparameters as the production Model B, so that the *only* difference between them is which feature subset they saw.
14. As a model developer, I want a paired bootstrap (1,000 resamples, 95% CI) over the same Testset sample pairs comparing the three ablation variants, so that I have a statistically defensible answer to "do the physical features actually help, beyond sampling noise."
15. As a model developer, I want Tree SHAP computed over all 8,964 Model B features with the top-20 sequence-embedding features and all 4 physical features reported separately, so that physical-feature contribution isn't buried inside a sequence-embedding-dominated ranking.
16. As a model developer, I want an aggregate physical SHAP contribution ratio (sum of |SHAP| over the 4 physical features, as a share of total |SHAP|), so that I have one headline number for "how much of Model B's decision-making is physical-feature-driven."
17. As a model developer, I want SHAP dependence plots (feature value vs. SHAP value) for each of the 4 physical features, so that I can sanity-check whether the learned feature-response relationship is at least qualitatively consistent with known CRISPR molecular biology.
18. As a model developer, I want Attention Rollout computed with the fixed operation order (rollout matrix = 0.5*A + 0.5*I, applied *before* row-normalization, across all NT transformer layers), so that every attention analysis in this pipeline uses one unambiguous, reproducible recipe rather than an equally-plausible alternative ordering.
19. As a model developer, I want the `[CLS]` token stripped and the remaining token-token attention matrix mapped back to 36bp nucleotide coordinates, so that the result is directly comparable to the ISM/IG position axes.
20. As a model developer, I want the Attention Rollout output explicitly labeled "Information Dependency" rather than described as physical DNA/protein binding, so that no one downstream mistakes an internal-attention artifact for wet-lab binding evidence.
21. As a researcher, I want Case Study samples selected by the documented residual-quantile rules (Primary Discordant, Secondary Discordant fallback, Concordant) with the exact fallback order (Primary first, Secondary fills remaining slots only if Primary yields fewer than 5), so that the 10 representative cases are picked by a reproducible rule, not manual curation.
22. As a researcher, I want each Case Study's entry to bundle its sequence, true score, both models' raw predictions and errors, physical feature values, per-sample physical SHAP with a stated interpretation, IG attribution arrays, and attention-rollout matrix, so that one JSON object gives a complete, self-contained explanation of that sample.
23. As a researcher, I want performance metrics (rho, r, MAE, MSE) computed on models' raw prediction scale with no cross-model rescaling, so that comparisons reflect true model behavior rather than an analyst's calibration choice; any scale normalization is confined to visualization only.
24. As a downstream dashboard/pipeline consumer, I want large multi-dimensional arrays (ism_delta, mismatch arrays, attention matrices) stored in a separate `model_analysis_arrays.npz`, and only metadata + the 10 Case Studies in `model_analysis_summary.json`, so that the dashboard's primary data file stays small and human-inspectable while full data stays available.
25. As a downstream dashboard/pipeline consumer, I want `model_analysis_summary.json` to follow the exact schema documented in `plan_realize.md` section 4 (metadata, array axis definitions, global_evaluation, complex_mismatch_metadata, case_studies), so that the dashboard can bind to a stable, known contract rather than reverse-engineering ad hoc output.
26. As a future agent maintaining this pipeline, I want each step's entry point testable purely against its documented output artifact (shape, value constraints, schema), so that I can verify correctness without mocking internals of transformers/XGBoost/ViennaRNA.
27. As a future agent maintaining this pipeline, I want Model B's XAI-only needs (gradients through the regression head, `output_attentions=True`) isolated in a dedicated `Model_B_XAIPredictor` class per ADR-0001, so that adding analysis capability never risks correctness or performance of the production `Model_B_Predictor.predict()` path.

## Implementation Decisions

- **New module: `Model_B_XAIPredictor`** (per ADR-0001, not yet implemented). Loads the same NT checkpoint path as `Model_B_Predictor` but via `AutoModelForSequenceClassification` (not `AutoModelForMaskedLM`), exposes a differentiable forward pass through the regression head and supports `output_attentions=True`. Feeds both Step 2 (Integrated Gradients) and Step 4 (4) (Attention Rollout). Does not modify `Model_B_Predictor`.
- **Model A IG**: computed directly against the existing `IntegratedPredictor` (`model_a_wrapper.py`) since it's already a plain differentiable `nn.Module`; baseline is an all-zero one-hot tensor.
- **Model B IG**: target is `Model_B_XAIPredictor`'s `(batch, 7, 1280)` token embedding tensor, asserted/validated before attribution runs; baseline is a fixed zero embedding vector of the same shape. If the shape assertion fails, the run must raise and stop — no silent fallback to a different layer.
- **6-mer to nucleotide projection**: Deterministic Projection Rule — each 6-mer token's attribution divided evenly across its 6 nucleotides (see `CONTEXT.md`). Must be documented in output as an analysis assumption, not a model-internals claim.
- **Attribution comparison basis**: within-sequence L1-normalized attribution, not raw magnitude; comparison is relative distribution (position-wise, especially PAM `NNGRRT` and Seed 1-8bp), not absolute value.
- **ISM**: 36 positions x 3 alternative bases = 108 mutants per sample; alternative-base axis fixed to lexicographic order (`sorted({A,C,G,T} - {WT})`). Output array `ism_delta`: shape `(514, 36, 3)`. Model B recomputes its 4 physical features per mutant (no reuse of wild-type physical features). Relative delta-Score masked to NaN where `|Score_WT| <= 1e-4`.
- **Mismatch profiling**: `mismatch_single_delta` (514, 36, 3) preserves all 3 substitutions per position; `mismatch_single_min` (514, 36) is the single worst-case summary. `mismatch_complex` (514, 15): 15 fixed scenarios (5 Seed + 5 Distal + 5 Intermittent), `random_seed=42`, each carrying `mutation_id`, `positions`, `ref`/`alt` bases, `region` metadata.
- **Ablation variants**: `Model B-Embedding` (8,960d), `Model B-Physical` (4d), `Model B-Full` (8,964d) — each independently retrained on the same dev split, same seed, and the same XGBoost hyperparameters already used in `nt_feature_stacking.py` (`n_estimators=500, learning_rate=0.05, max_depth=6, subsample=0.8, colsample_bytree=0.8, early_stopping_rounds=20`). Compared via paired bootstrap (1,000 resamples, 95% CI) on the same Testset sample pairs.
- **Tree SHAP**: computed over all 8,964 features; report top-20 embedding features and all 4 physical features separately, plus an aggregate physical SHAP contribution ratio (sum|SHAP_physical| / sum|SHAP_all|).
- **Attention Rollout operation order** (fixed, applies to every computation in this pipeline): `rollout = 0.5*A + 0.5*I` applied first, row-normalization applied second. `[CLS]` removed before mapping token coordinates back to the 36bp nucleotide coordinate system. Result is documented as "Information Dependency," explicitly not a physical binding claim.
- **Case Study selection** (residuals `e_A = |y - y_hat_A|`, `e_B = |y - y_hat_B|`):
  - Primary Discordant: `e_A >= Q75(e_A) AND e_B <= Q25(e_B)` — top 5 preferred.
  - Secondary Discordant (fallback only): `(e_A - e_B)/e_A >= 0.5 AND e_A >= Median(e_A)` — used to fill remaining slots only if Primary yields fewer than 5.
  - Concordant: `e_A <= Q25(e_A) AND e_B <= Q25(e_B)` — top 5.
  - Total Case Study sample cap: 10 (5 Discordant-family + 5 Concordant).
- **No cross-model score rescaling**: performance metrics (rho, r, MAE, MSE) always computed on raw prediction scale for both models; any scale normalization is visualization-only and never feeds back into metric computation.
- **Export schema** — two artifacts:
  - `model_analysis_arrays.npz`: full (514, ...) tensors — `ism_delta`, `mismatch_single_delta`, `mismatch_single_min`, `mismatch_complex`, attention matrices, per-sample IG attribution arrays.
  - `model_analysis_summary.json`: metadata (testset version, sample count, SHA256, Model B architecture dims, array axis definitions), `global_evaluation` (ablation results + paired bootstrap CI + aggregate physical SHAP ratio), `complex_mismatch_metadata`, and the 10 `case_studies` entries. Exact field shape (from `plan_realize.md`, itself acting as the schema-defining prototype for this decision):
    ```json
    {
      "metadata": { "testset_version": "...", "total_samples": 514, "sha256_checksum": "...",
        "model_b_architecture": { "tokenizer": "6-mer", "tokens": 7, "hidden_dim": 1280,
          "flattened_dim": 8960, "physical_dim": 4, "total_feature_dim": 8964 },
        "array_axis_definitions": { "...": "..." } },
      "global_evaluation": { "ablation_results": { "embedding_only": {}, "physical_only": {},
          "full_model": {}, "paired_bootstrap_95ci": { "delta_spearman": [], "delta_mse": [] } },
        "aggregate_physical_shap_contribution_ratio": 0.0842 },
      "complex_mismatch_metadata": [ { "scenario_index": 0, "mutation_id": "...", "region": "Seed",
          "positions": [1,2], "description": "..." } ],
      "case_studies": [ { "case_id": "DISCORDANT_P01", "sample_id": 42, "original_id": 418,
          "sequence": "...", "true_score": 0.82, "model_a": {}, "model_b": {},
          "case_type": "Primary Discordant", "per_sample_physical_values": {},
          "per_sample_physical_shap": {"interpretation": "..."},
          "integrated_gradients": {}, "attention_rollout": {} } ]
    }
    ```

## Testing Decisions

- Test each step's **top-level output artifact**, not internal calls into `transformers`/`xgboost`/`RNA`/`Bio.SeqUtils` — a good test here checks shape, dtype, value-range/gating behavior, and schema conformance of what the step returns or writes, not how it got there.
- **Step 2 seam**: given Testset sequences + both predictors, assert per-sample 36bp normalized attribution arrays for Model A and Model B — correct shape, L1-normalized rows, NaN/error on a deliberately malformed Model B embedding shape (validates the pre-flight assertion), and correct even-split values under the Deterministic Projection Rule for a synthetic single-token case.
- **Step 3 seam**: assert `ism_delta` shape `(N, 36, 3)` with the fixed lexicographic alt-base order, NaN masking at the `|Score_WT| <= 1e-4` boundary, `mismatch_single_min` equal to the min over `mismatch_single_delta`'s last axis, and `mismatch_complex` producing exactly 15 scenarios with the documented region split and reproducible under `seed=42`.
- **Step 4 seam**: assert the three ablation variants are trained on the documented shared protocol and that paired bootstrap CI bounds are computed over matched sample pairs; assert Tree SHAP output separates top-20 embedding vs. the 4 named physical features and that the aggregate physical ratio is a valid sum|SHAP| proportion; assert Attention Rollout applies `0.5*A + 0.5*I` before row-normalization (not after) on a small synthetic attention matrix where the two orderings produce numerically distinguishable results, and that `[CLS]` is excluded from the final `(36, 36)` matrix.
- **Step 5 seam**: assert Case Study selection reproduces the documented Primary -> Secondary-fallback -> Concordant rule (including the "fewer than 5 Primary" fallback trigger) on a synthetic residual distribution with a known expected selection, and assert `model_analysis_summary.json` output validates against the Section-4 schema (required keys present, `case_studies` length <= 10, arrays absent from the JSON and present in the `.npz` instead).
- **Prior art**: `step1_verify.py` is this repo's existing example of black-box, output-focused verification (load Testset -> run both predictors -> check metrics) — the pattern to follow, extended to each new step's own artifact.
- Modules under test: the new `Model_B_XAIPredictor`, and one function/entry point per Step 2-5 producing its documented artifact fragment. `Model_A_Predictor`/`Model_B_Predictor` are exercised as fixtures (already covered by Step 1), not re-tested here.

## Out of Scope

- Retraining or architecturally changing Model A.
- Building the XAI demo dashboard UI itself (project "C") — this spec only produces the data it will consume.
- Wet-lab or experimental biological validation of any hypothesis surfaced by this analysis.
- Performance/runtime optimization of the analysis pipeline (e.g. GPU batching strategy, RNAfold/RNAduplex call batching).
- Any dataset beyond the existing 514-sample, SHA256-verified Testset.
- Changing `Model_B_Predictor`'s production inference path or its hyperparameters.

## Further Notes

- Step 4-1's three ablation retrains and Step 4-2's Tree SHAP over 8,964 features are the most compute-heavy pieces of this spec; expect them to dominate runtime.
- `_calc_mfe`/`_calc_duplex_dg` in `model_b_wrapper.py` shell out to the `RNAfold`/`RNAduplex` CLI (ViennaRNA); Step 3's per-mutant physical-feature recomputation (108 mutants x 514 samples) multiplies that CLI-call cost substantially — worth flagging even though optimization itself is out of scope.
- ADR-0001 is the load-bearing architectural decision this spec builds on (`Model_B_XAIPredictor` split from `Model_B_Predictor`); if a future implementer is tempted to add XAI methods directly onto `Model_B_Predictor` instead, that contradicts ADR-0001 and should be flagged rather than silently done.
